### DISCUSSION TITLE
Num_workers for GPU

### DISCUSSION THREAD

**Discussion by rlsandovalp** [neuralhydrology/neuralhydrology#127] [Q&A] [answered]
Hi, 

I'm new in neural hydrology and I'm wondering why the training time of the Introduction Notebook increases when I increase the number of workers and decreases a lot when I set it to 0 (I'm using a gpu). I _guess_ that is a feature that has a lot of sense when using cpu but not so much when using gpu, am I wrong? 

Thanks for your help :)

**Comment by gauchm** [comment, marked as answer]
Actually, in principle, `num_workers` can be (and usually is) useful both for GPU and CPU training. The argument controls whether minibatches are loaded from the dataset in the main process which also runs the training loop (`num_workers = 0`) or whether PyTorch spawns separate processes that are responsible for prefetching the minibatches (`num_workers > 0`).

For the introduction notebook specifically, I think the reason why more workers make things worse is that the dataset is so tiny and GPUs are so fast that the overhead of spawning the additional workers outweighs the long-term benefits of prefetched minibatches.
On a larger training set (like training on the full CAMELS dataset), num_workers > 0 should improve your training speed, even on GPU.

In case you're interested, you can read up on the details of Pytorch dataloading and num_workers [here](https://pytorch.org/docs/stable/data.html#single-and-multi-process-data-loading).

**Reply by rlsandovalp** [reply]
Thanks
